### ISSUE TITLE
starlark: use less memory

### ISSUE THREAD

**Issue by alandonovan** [google/starlark-go#184]
% alias maxrss='command time -f %MKB'

% maxrss starlark -c 'list(range(10000000))'
339240KB

That's an average of 34 bytes per integer list element:
2 words for the Value reference in the list
2 words for the Int struct (all these values are small)
6% malloc overhead, which is not bad
= 34 bytes

% maxrss starlark -c '[""] * 100000000'
1624056KB
A list of duplicate empty strings uses 16 bytes per list element.
2 words for the Value
(+ 1 shared convTstring alloc)
= 16 bytes

% maxrss starlark -c '[""[:] for x in range(100000000)]'
maxrss 6947044KB
A list of distinct empty strings uses 69 bytes per list element.
2 words for the Value
2 words for each string convTstring alloc
 + 4.6 words overhead (malloc? appends during list comprehension?)

Even worse:
% maxrss starlark -c '["%s" % "" for x in range(100000000)]'
9335208KB
A list of distinct (empty) strings has 93 bytes per element.
2 words for the Value
2 words for each string convTstring alloc
3 words for each string allocated by %s (?)
+ 4.6 words overhead as before

**Comment by josharian** [comment]
Profiling (via #190) of `% maxrss starlark -c '["%s" % "" for x in range(100000000)]'` yields

```
)$ go tool pprof -alloc_space -lines `which starlark` m.p 
File: starlark
Type: alloc_space
Time: Apr 11, 2019 at 10:08am (PDT)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top5 
Showing nodes accounting for 13.47GB, 100% of 13.47GB total
Showing top 5 nodes out of 13
      flat  flat%   sum%        cum   cum%
    8.97GB 66.59% 66.59%     8.97GB 66.59%  go.starlark.net/starlark.(*Function).CallInternal /Users/josh/src/go.starlark.net/starlark/interp.go:400
    2.99GB 22.22% 88.82%     2.99GB 22.22%  go.starlark.net/starlark.interpolate /Users/josh/src/go.starlark.net/starlark/eval.go:1319
    1.51GB 11.18%   100%     1.51GB 11.18%  go.starlark.net/starlark.rangeValue.Index /Users/josh/src/go.starlark.net/starlark/library.go:895
         0     0%   100%     2.99GB 22.22%  go.starlark.net/starlark.(*Function).CallInternal /Users/josh/src/go.starlark.net/starlark/interp.go:147
         0     0%   100%     1.51GB 11.18%  go.starlark.net/starlark.(*Function).CallInternal /Users/josh/src/go.starlark.net/starlark/interp.go:317
```

The majority of the allocations are from growing the stack. It is the last line of this case in the interpreter:

```go
		case compile.APPEND:
			elem := stack[sp-1]
			list := stack[sp-2].(*List)
			sp -= 2
			list.elems = append(list.elems, elem)
```

I haven't dug enough yet to understand why this is so. Is iteration implemented as recursion?

**Comment by alandonovan** [comment]
Thanks Josh.  Append is the main operation done by [x for y in z], so that's not surprising. One could reserve len(z) space for a list comprehension, but that requires that you know len(z)---and z may be iterable but not of known length---and that there are no 'if' clauses in the comprehension.

In a larger benchmark I find that NewFrame is the hottest alloc, because every function call allocates one. We could probably save a slice of frames in the thread and grow it in larger chunks. I would like to remove the NewFrame function---there is no longer any need for clients to access it.

Thanks for -alloc_space, I wasn't aware of that. All this time I have been looking at in-use profiles.
One could fill a book (or free wiki) with useful information about program dynamics and profiling.

**Comment by josharian** [comment]
> Thanks for -alloc_space, I wasn't aware of that.

Yeah. IMHO, it should be the default when reading from a file, and in-use should be the default when reading over the network. There's some kind of pprof support for setting the default, but I don't think runtime/pprof exposes it. It's all a bit of a mess. Oh, and just in case, there's also `-alloc_objects`. :)

**Comment by alandonovan** [comment]
Some observations from profiling a substantial benchmark:

Two optimizations reduce garbage generation significantly: avoiding most allocations in NewFrame by saving Frames in a per-Thread freelist, and avoiding most calls to stack=make([]Value) in Function.CallInternal by allocating space for 8 Values in the Frame. The profile for NewFrame+CallInternal goes down from 19% of total allocs to 12%. These optimizations are not sound with the current spaghetti-stack Frame design, so I propose to switch to using slices.

In MAKELIST, NewList accounts for 912MB whereas the elements account for only 714MB. In other words, most lists are so short that the header dominates the cost. We should use inline allocation of one or two elements.

In MAKETUPLE, the convT operation is as expensive as the make([]Value) of the elements. Unfortunately there's no way to inline allocate the convT (it can easily be done with unsafe but it really is unsafe). We should at least special case zero-length tuples to avoid the convT alloc.
